src/NumericalAlgorithms/RootFinding/GslMultiRoot.cpp

```diff
@@ -0,0 +1,151 @@
+
+#include <exception>
+#include <gsl/gsl_multiroots.h>
+#include <gsl/gsl_vector.h>
+
+#include "ErrorHandling/Assert.hpp"
+#include "ErrorHandling/Error.hpp"
+#include "Parallel/Printf.hpp"
+
+class rootfinding_exception : public std::exception {
+  virtual const char* what() const throw() { return "GSLMultiRoot failed."; }
+};
+
+void printstate_GSLMultiRootFinder(size_t iter, gsl_multiroot_fdfsolver* s,
+                                   size_t dim, bool PrintHeader = false) {
+  if (PrintHeader){
+    Parallel::printf("Iter\t");
+    for (size_t i = 0; i < dim; ++i) {
+      Parallel::printf(" x[%u]\t", i);
+    }
+    for (size_t i = 0; i < dim; ++i) {
+      Parallel::printf(" f[%u]\t", i);
+    }
+    Parallel::printf("\n");
+  }
+
+  Parallel::printf("%u\t", iter);
+  for (size_t i = 0; i < dim; ++i) {
+    Parallel::printf("%3.4f  ", gsl_vector_get(s->x, i));
+  }
+  for (size_t i = 0; i < dim; ++i) {
+    Parallel::printf("%1.3e  ", gsl_vector_get(s->f, i));
+  }
+  Parallel::printf("\n");
+}
+
+template <size_t Dim>
+std::array<double, Dim> GSLMultiRoot(
+    const GslMultiRootFunctionFdf<Dim>& func,
+    const std::array<double, Dim> initial_guess,
+    const double& absolute_tolerance, const size_t maximum_iterations,
+    const double relative_tolerance, const Verbosity verbosity,
+    const double maximum_absolute_tolerance, const Method method) {
+  gsl_multiroot_function_fdf f = {
+      &Gsl_MultiRootFunctionFdf_Wrapper_f<Dim>,
+      &Gsl_MultiRootFunctionFdf_Wrapper_df<Dim>,
+      &Gsl_MultiRootFunctionFdf_Wrapper_fdf<Dim>, Dim,
+      const_cast<GslMultiRootFunctionFdf<Dim>*>(&func)};
+
+  // Set up initial guess as stored in initial_guess:
+  gsl_vector* gsl_root = gsl_vector_alloc(Dim);
+  for (size_t i = 0; i < Dim; ++i) {
+    gsl_vector_set(gsl_root, i, initial_guess[i]);
+  }
+
+  // Set up method for solver:
+  const gsl_multiroot_fdfsolver_type* solver_type;
+  gsl_multiroot_fdfsolver* solver;
+  if (method == Method::GNewton) {
+    solver_type = gsl_multiroot_fdfsolver_gnewton;
+  } else if (method == Method::Hybridsj) {
+    solver_type = gsl_multiroot_fdfsolver_hybridsj;
+  } else if (method == Method::Hybridj) {
+    solver_type = gsl_multiroot_fdfsolver_hybridj;
+  } else {
+    ERROR(
+        "Invalid method. Has to be one of GNewton, Hybridsj or "
+        "Hybridj.");
+  }
+  solver = gsl_multiroot_fdfsolver_alloc(solver_type, Dim);
+  gsl_multiroot_fdfsolver_set(solver, &f, gsl_root);
+
+  // Take iterations:
+  int status;
+  size_t iter = 0;
+  do {
+    iter++;
+    status = gsl_multiroot_fdfsolver_iterate(solver);
+    // Check if solver is stuck
+    if (status == GSL_ENOPROG) {
+      break;
+    }
+
+    // Check termination criteria depending on whether or not relative_tolerance
+    // is specified
+    if (relative_tolerance > 0) {
+      status = gsl_multiroot_test_delta(solver->dx, solver->x,
+                                        absolute_tolerance, relative_tolerance);
+    } else {
```

Comment:
The documentation should explain that `relative_tolerance <= 0` has a special meaning.

Reply:
Should I instead assert on having a non-negative relative_tolerance? This chunk of code isn't so much that something special is done with a negative value of relative_tolerance since it gets ignored entirely if it's zero or below

Reply:
But it appears that `relative_tolerance = 0` does something very different from `relative_tolerance = epsilon`.

Reply:
I understand what you mean now